tests/test_upload.py

```diff
@@ -117,6 +89,39 @@ def test_deprecated_repo(make_settings):
     )
 
 
+def test_exception_for_redirect(make_settings):
+    upload_settings = make_settings("""
+        [pypi]
+        repository: https://test.pypi.org/legacy
+        username:foo
+        password:bar
+    """)
+
+    stub_response = pretend.stub(
+        is_redirect=True,
+        status_code=301,
+        headers={'location': 'https://test.pypi.org/legacy/'}
+    )
+
+    stub_repository = pretend.stub(
+        upload=lambda package: stub_response,
+        close=lambda: None
+    )
+
+    upload_settings.create_repository = lambda: stub_repository
+
+    with pytest.raises(exceptions.RedirectDetected) as err:
+        upload.upload(upload_settings, [WHEEL_FIXTURE])
+
+    assert err.value.args[0] == (
+        "https://test.pypi.org/legacy attempted to redirect"
```

Comment:
This is something I struggle with often. As a CLI, our user interface is what we print out. With that in mind, tests like these quickly become frustrating and feel overly strict when updating messages/message formatting/etc. I think testing that we're showing the correct data is important, but I also don't want us to test that we always show this message. I can imagine a future where we use internationalization to improve the experience and someone contributing from a locale that isn't `en_*` sees a test failure here too. 

What do you think are the important parts of this message? What should we test for and what can we eliminate from this assertion?

Reply:
Thanks, @sigmavirus24. In this case, I followed the pattern of the existing tests, but in generally I agree with your opinion. To that end, in f326758 I modified the assertions to check for what I think are the important bits, like URLs.